File: ruby/affine-cipher/affine_cipher.rb
class Integer
  ## find `n` where `a*n mod m == 1`
  def modular_multiplicative_inverse(m)
    raise ArgumentError, "#{self} and #{m} are not coprime" unless gcd(m) == 1

    # this simple approach is brute force, not the most efficient.
    n = 0
    n += 1 until send(:*, n).modulo(m) == 1
    n
  end
end

class Affine
  ALPHABET = ('a'..'z').to_a.join('').freeze

  def initialize(a, b)
    @mmi = a.modular_multiplicative_inverse ALPHABET.length
    @coefficients = [a, b]
  end

  def encode(plaintext)
    encoded = xxcode(plaintext) { |x| E(x) }
    encoded.scan(/.{1,5}/).join(' ')
  end

  def decode(ciphertext)
    xxcode(ciphertext) { |x| D(x) }
  end

  private

  def xxcode(text, &block)
    t = text.downcase.gsub(/[^[:alnum:]]/, '')
    t.chars.map(&block).join('')
  end

  def E(char)
    x = ALPHABET.index char
    return char if x.nil?

    a, b = @coefficients
    m = ALPHABET.length
    ALPHABET[ (a * x + b) % m ]
  end

  def D(char)
    y = ALPHABET.index char
    return char if y.nil?

    _a, b = @coefficients
    m = ALPHABET.length
    a_inv = @mmi
    ALPHABET[ (a_inv * (y - b)) % m ]
  end
end

File: python/affine-cipher/affine_cipher.py
from string import ascii_lowercase as alphabet, digits
from math import gcd
import re

non_alnum = f'[^{alphabet}{digits}]'


def encode(plain_text, a, b):
    validate_key(a)

    def encoding(char):
        try:
            x = alphabet.index(char)
            char = alphabet[(a * x + b) % len(alphabet)]
        except ValueError:
            pass
        return char

    encoded = x_code(plain_text, encoding)
    return re.sub('.{5}', r'\g<0> ', encoded).rstrip(" ")


def decode(ciphered_text, a, b):
    validate_key(a)
    a_inv = mmi(a, len(alphabet))

    def decoding(char):
        try:
            y = alphabet.index(char)
            char = alphabet[(a_inv * (y - b)) % len(alphabet)]
        except ValueError:
            pass
        return char

    return x_code(ciphered_text, decoding)


def validate_key(a):
    if gcd(a, len(alphabet)) != 1:
        raise ValueError('a and m must be coprime.')


def mmi(a, m):
    for n in range(m):
        if (a * n) % m == 1:
            return n


def x_code(text, func):
    return ''.join(
        func(char)
        for char in re.sub(non_alnum, '', text.lower())
    )
